### DISCUSSION TITLE
sdm vs pi-gen

### DISCUSSION THREAD

**Discussion by alexjaw** [gitbls/sdm#46] [General]
Since sdm and [pi-gen](https://github.com/RPi-Distro/pi-gen) both can be used to configure Pi OS images, I wonder if you could high-light the advantages of using sdm?

Keep up the great work!
Alex

**Comment by gitbls** [comment]
Great question! Here's my take on these, with a look at Pi Imager as well. If I've duffed or under-explained anything, I'd very much appreciate the feedback.

Here's  a brief overview of the three tools. Detailed pros/cons follow

* pi-gen (https://github.com/RPi-Distro/pi-gen) is a low-level tool that is used for building RasPiOS IMG files. It enables you do pretty much everything, since it’s the tool used by the RasPiOS team to build the various IMGs they distribute. It's moderately complex to set up, but provides really fine-grain control over what is included in the IMG

* sdm (https://github.com/gitbls/sdm) is a mid-level tool. It assumes the output of pi-gen as its input (starts with the IMGs that RPF distributes). Relatively easy to use, since it’s command-line based. You can actually do quite a bit without writing a single bash script or config file, but obviously can do a lot more as needed/desired. Sdm can be used to easily create your own customized IMG file, and then burn to SSD/SD, doing any final per-target system customization after burning. It also easily removes the first system re-boot by expanding the root partition as part of the burn process.

* Pi Imager (https://www.raspberrypi.com/software/), from RPT is a GUI that enables some customizations, in general the important ones for casual/GUI-oriented users. But it falls short of both pi-gen and sdm in terms of overall depth of functionality. It can also handle IMGs from other distros. sdm, on the other hand, is more RasPiOS focused.

* IMGs produced with pi-gen can be further customized by sdm and burned using either sdm or pi imager
* IMGs customized by sdm can be burned using either sdm or pi imager

More detailed pros/cons of each:

* **pi-gen**
  * Pros
    * Fine-grained control over the contents of the IMG. IMG can contain exactly the packages you want to start with, and how they are configured
    * Supports multiple distros and architectures
  * Cons
    * Relatively complex setup; no installer, and 32-vs-64 bit install differences
    * Documentation lacking some detail, dependent on user expertise
    * Must edit possibly multiple scripts to do just about anything
    * Bash knowledge required
    * The resulting IMG is likely unique. Nobody attempting to help you will have exactly the same IMG
    * Doesn't help at all with burning the IMG to an SSD/SD, or with easily producing customized target disks from a single IMG

* **sdm**
  * Pros
    * Relatively simple installer; Installer provided
    * Starts with an already-produced IMG (e.g. official RasPiOS IMGs using pi-gen) and extends it via automation
    * Simple command-line interface that can be scripted
    * Complete, detailed documentation
    * All customizations are done using standard, well-defined tools and interfaces, such as raspi-config, systemd, etc.
    * Sigificant customization can be done without editing any scripts; complex customizations are relatively simple with well-documented techniques
    * Easy to extend via structured plug-in capability
    * Easy to burn multiple uniquely-customized SSD/SD cards from same IMG
    * Includes capability to explore an IMG/SSD/SD in a contained environment
    * Can customize both 32 and 64-bit IMGs on both 32 and 64-bit hosts
    * Bash knowledge not required for basic-to-intermediate use, but some knowledge helpful for advanced techniques
  * Cons
    * Only supports RasPiOS. Others may work, but are completely untested
    * More coarse-grained control over the IMG contents than pi-gen
    * IMG must be downloaded and uncompressed
    * Command-line, rather than config-file driven. some prefer config file due to richer expression

* **pi imager**
  * Pros
    * Simple installation/setup
    * Easy to use GUI
    * Can download IMG (even compressed) or burn from local copy
    * Enables a few simple customizations such as add user, WiFi configuration, SSH key, locale, hostname
    * Can burn IMGs for other Raspberry Pi distros
    * Includes tools to test/format SD Card
  * Cons
    * Literally no control over the IMG contents
